### PR TITLE
syslog-format: cannot parse ipv6 into hostname

### PR DESCRIPTION
**Pull request by furiel**
Fixes:
https://github.com/balabit/syslog-ng/issues/1616

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build FAILURE, the tests were executed on test branch: master and test suite: functions

**Comment by furiel** [comment]
Not nice but maybe works:

We could use inet_pton to check if we see a valid ipv6 address. If yes, we set the hostname. If not, we go with the original heuristic.

Of course we need to prepare the candidate string for inet pton, so first I need to create nullterminated substring until the first space or '[' character, which then can be passed to inet_pton.

**Comment by furiel** [comment]
Or this is maybe better: g_hostname_is_ip_address (), from glib 2.22

**Comment by bazsi** [comment]
never forget the performance implications, this is a very hot path here.


-- 
Bazsi

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by furiel** [comment]
Instead of explicitely validating for ipv6 address, I added a new version which relaxes the condition for colon being delimiter. The new heuristics just check a few extra characters around the colon, so probably have lower effect on performance than calling a glib function or inet_pton.

**Comment by presidento** [comment]
We should check a bit more strictly that it seems like an ipv6 address. (Hexa numbers, colons...)

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by furiel** [comment]
@bazsi I replaced the previous heuristics with a more strict one. In the end, an ipv6-heuristics class added just checks if the candidate is max 8 blocks of max 4-wide hexa characters. I felt the class useful, because the state of the ipv6 part became more complex than the previous one.

The good news is I could still avoid memory allocation, so I am very hopeful that we do not lose too much performance. Anyway, we will run a performance test for this version next week.

File: modules/syslogformat/syslog-format.c [hunk 1]
```diff
@@ -635,6 +635,48 @@ _init_parse_hostname_invalid_chars(void)
...
+        {
+          self->heuristic_failed = TRUE;
+          return FALSE;
+        }
+    }
+
+  if (c == ':')
+    {
+      self->digits_in_segment = 0;
+      if (self->current_segment > 8)
```
Comment (MrAnno) [inline modules/syslogformat/syslog-format.c]: `current_segment` is not incremented.

File: modules/syslogformat/syslog-format.c [hunk 2]
```diff
@@ -635,6 +635,48 @@ _init_parse_hostname_invalid_chars(void)
     }
 }
 
+typedef struct _IPv6Heuristics
+{
+  gchar current_segment;
+  gchar digits_in_segment;
```
Comment (MrAnno) [inline modules/syslogformat/syslog-format.c]: Please use `guint8` for 8 bit integers.

File: modules/syslogformat/syslog-format.c [hunk 3]
```diff
@@ -635,6 +635,48 @@ _init_parse_hostname_invalid_chars(void)
...
+  gboolean heuristic_failed;
+  gboolean (*feed_gchar)(struct _IPv6Heuristics *, gchar);
```
Comment (MrAnno) [inline modules/syslogformat/syslog-format.c]: Is there any reason why you made this virtual?

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by pzoleex** [comment]
the perftest passed, there was no regression

**Pull request merged**